Failure to build with Boost 1.77

Build testing the OpenBSD ports tree with Boost 1.77 beta1. bitcoin 0.21.1 fails to build as shown..

```
wallet/test/init_test_fixture.cpp:34:52: error: no member named 'BOOST_FILESYSTEM_C_STR' in 'boost::filesystem::path'
    std::ofstream f(m_walletdir_path_cases["file"].BOOST_FILESYSTEM_C_STR);
                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
1 error generated.
```

Complete noob here, sorry, but is this related?:

On Centos 7, with
Package boost-1.53.0-28.el7.x86_64 already installed and latest version
Package boost-devel-1.53.0-28.el7.x86_64 already installed and latest version
```
In file included from /usr/include/boost/optional.hpp:15:0,
                 from ./optional.h:10,
                 from ./util/system.h:22,
                 from ./addrman.h:17,
                 from ./net.h:10,
                 from net.cpp:10:
/usr/include/boost/optional/optional.hpp: In instantiation of ‘void boost::optional_detail::optional_base<T>::construct(const Expr&, const void*) [with Expr = CDataStream; T = CNetMessage]’:
/usr/include/boost/optional/optional.hpp:262:25:   required from ‘boost::optional_detail::optional_base<T>::optional_base(const Expr&, const Expr*) [with Expr = CDataStream; T = CNetMessage]’
/usr/include/boost/optional/optional.hpp:559:78:   required from ‘boost::optional<T>::optional(const Expr&) [with Expr = CDataStream; T = CNetMessage]’
net.cpp:761:47:   required from here
/usr/include/boost/optional/optional.hpp:392:8: error: no matching function for call to ‘CNetMessage::CNetMessage(const CDataStream&)’
        new (m_storage.address()) internal_type(expr) ;
        ^
/usr/include/boost/optional/optional.hpp:392:8: note: candidates are:
In file included from net.cpp:10:0:
./net.h:742:5: note: CNetMessage::CNetMessage(CDataStream&&)
     CNetMessage(CDataStream&& recv_in) : m_recv(std::move(recv_in)) {}
     ^
./net.h:742:5: note:   no known conversion for argument 1 from ‘const CDataStream’ to ‘CDataStream&&’
./net.h:734:7: note: CNetMessage::CNetMessage(const CNetMessage&)
 class CNetMessage {
       ^
./net.h:734:7: note:   no known conversion for argument 1 from ‘const CDataStream’ to ‘const CNetMessage&’
./net.h:734:7: note: CNetMessage::CNetMessage(CNetMessage&&)
./net.h:734:7: note:   no known conversion for argument 1 from ‘const CDataStream’ to ‘CNetMessage&&’
make[2]: *** [libbitcoin_server_a-net.o] Error 1
make[2]: Leaving directory `/home/pjalajas/Documents/dev/hub/test/projects/bitcoincore/bitcoin-0.21.1/src'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/home/pjalajas/Documents/dev/hub/test/projects/bitcoincore/bitcoin-0.21.1/src'
make: *** [all-recursive] Error 1
```

With the following diff
```diff
--- a/depends/packages/boost.mk
+++ b/depends/packages/boost.mk
@@ -1,8 +1,8 @@
 package=boost
-$(package)_version=1_71_0
+$(package)_version=1_77_0
 $(package)_download_path=https://boostorg.jfrog.io/artifactory/main/release/$(subst _,.,$($(package)_version))/source/
 $(package)_file_name=boost_$($(package)_version).tar.bz2
-$(package)_sha256_hash=d73a8da01e8bf8c7eda40b4c84915071a8c8a0df4a6734537ddde4a8580524ee
+$(package)_sha256_hash=fc9f85fc030e233142908241af7a846e60630aa7388de9a5fafb1f3a26840854
 $(package)_dependencies=native_b2
 
 define $(package)_set_vars
```

I can confirm the issue:
```
wallet/test/init_test_fixture.cpp: In constructor ‘InitWalletDirTestingSetup::InitWalletDirTestingSetup(const string&)’:
wallet/test/init_test_fixture.cpp:35:52: error: ‘std::map<std::__cxx11::basic_string<char>, boost::filesystem::path>::mapped_type’ {aka ‘class boost::filesystem::path’} has no member named ‘BOOST_FILESYSTEM_C_STR’
   35 |     std::ofstream f(m_walletdir_path_cases["file"].BOOST_FILESYSTEM_C_STR);
      |                                                    ^~~~~~~~~~~~~~~~~~~~~~
make[2]: *** [Makefile:14782: wallet/test/test_test_bitcoin-init_test_fixture.o] Error 1
```

--- a/src/fs.cpp
+++ b/src/fs.cpp
@@ -242,7 +242,11 @@ void ofstream::close()
 }
 #else // __GLIBCXX__
 
+#if BOOST_VERSION >= 107700
+static_assert(sizeof(*BOOST_FILESYSTEM_C_STR(fs::path())) == sizeof(wchar_t),
+#else
 static_assert(sizeof(*fs::path().BOOST_FILESYSTEM_C_STR) == sizeof(wchar_t),
+#endif // BOOST_VERSION >= 107700
     "Warning: This build is using boost::filesystem ofstream and ifstream "
     "implementations which will fail to open paths containing multibyte "
     "characters. You should delete this static_assert to ignore this warning, "

--- a/src/wallet/test/db_tests.cpp
+++ b/src/wallet/test/db_tests.cpp
@@ -25,7 +25,11 @@ BOOST_AUTO_TEST_CASE(getwalletenv_file)
     std::string test_name = "test_name.dat";
     const fs::path datadir = gArgs.GetDataDirNet();
     fs::path file_path = datadir / test_name;
+#if BOOST_VERSION >= 107700
+    std::ofstream f(BOOST_FILESYSTEM_C_STR(file_path));
+#else
     std::ofstream f(file_path.BOOST_FILESYSTEM_C_STR);
+#endif // BOOST_VERSION >= 107700
     f.close();
 
     std::string filename;

--- a/src/wallet/test/init_test_fixture.cpp
+++ b/src/wallet/test/init_test_fixture.cpp
@@ -32,7 +32,11 @@ InitWalletDirTestingSetup::InitWalletDirTestingSetup(const std::string& chainNam
     fs::create_directories(m_walletdir_path_cases["default"]);
     fs::create_directories(m_walletdir_path_cases["custom"]);
     fs::create_directories(m_walletdir_path_cases["relative"]);
+#if BOOST_VERSION >= 107700
+    std::ofstream f(BOOST_FILESYSTEM_C_STR(m_walletdir_path_cases["file"]));
+#else
     std::ofstream f(m_walletdir_path_cases["file"].BOOST_FILESYSTEM_C_STR);
+#endif // BOOST_VERSION >= 107700
     f.close();
 }
 

